### ISSUE TITLE
NoSuchMethodError

### ISSUE THREAD

**Issue by dkambale** [twilio/twilio-java#603]
Exception in thread "main" java.lang.NoSuchMethodError: org.apache.http.impl.client.DefaultRedirectStrategy.<init>([Ljava/lang/String;)V
	at com.twilio.http.HttpClient.<init>(HttpClient.java:19)

**Comment by childish-sambino** [comment]
Could you provide more details, library/SDK versions, code samples, and links for what you're trying to accomplish?

**Comment by jagtejsodhi** [comment]
+1 here.

SDK 8.1.0

Trace:

```
        at com.twilio.http.HttpClient.<init>(HttpClient.java:19) ~[twilio-8.2.0.jar!/:na]
	at com.twilio.http.NetworkHttpClient.<init>(NetworkHttpClient.java:46) ~[twilio-8.2.0.jar!/:na]
	at com.twilio.http.NetworkHttpClient.<init>(NetworkHttpClient.java:35) ~[twilio-8.2.0.jar!/:na]
	at com.twilio.http.TwilioRestClient$Builder.build(TwilioRestClient.java:123) ~[twilio-8.2.0.jar!/:na]
	at com.twilio.Twilio.buildRestClient(Twilio.java:187) ~[twilio-8.2.0.jar!/:na]
	at com.twilio.Twilio.getRestClient(Twilio.java:163) ~[twilio-8.2.0.jar!/:na]
	at com.twilio.base.Creator.create(Creator.java:40) ~[twilio-8.2.0.jar!/:na]
```

**Comment by jagtejsodhi** [comment]
Hey @dkambale,

I found that SDK version 7.55.3 introduce apache's RedirectStrategy here: https://github.com/twilio/twilio-java/blob/3e1fa596cac9ffba7af84f15180acd65ebd9d0df/src/main/java/com/twilio/http/HttpClient.java#L19.

The constructor at that line is invalid, and thus throws a NoSuchMethodError. I reverted back to version 7.55.2 and it resolved my issue.

@childish-sambino please look into this as well. Thanks!

![123778996_369273430998778_5114827514865281374_n](https://user-images.githubusercontent.com/10988161/98426063-545e4680-204c-11eb-90db-98ae21e8cbdb.png)

**Comment by thinkingserious** [comment]
@jagtejsodhi,

I was unable to reproduce this locally using SDK version 8.2.0 with openjdk 11.0.9 2020-10-20. Could you provide a bit more detail so that I may try to reproduce? (e.g. java version)

Thanks!

**Comment by yungang-bgl** [comment]
I met the similar problem in my project. It might due to the incompatible `org.apache.httpcomponents.httpclient` runtime version. In Twilio SDK `httpclient` version is 4.5.12:compile, while in my project it's 4.0.1:runtime. I manually added the dependency then it works.
`        <dependency>
            <groupId>org.apache.httpcomponents</groupId>
            <artifactId>httpclient</artifactId>
            <version>4.5.12</version>
        </dependency>`

**Comment by JenniferMah** [comment]
@yungang-bgl 
Thanks for posting your solution! Do you mind including what version of java you were using?

**Comment by yungang-bgl** [comment]
Hi @JenniferMah , my Java version is
```
openjdk 11.0.4 2019-07-16 LTS
OpenJDK Runtime Environment Corretto-11.0.4.11.1 (build 11.0.4+11-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.4.11.1 (build 11.0.4+11-LTS, mixed mode)
```

**Comment by JenniferMah** [comment]
@dkambale
Are you still seeing the exception being thrown?

**Comment by yoovan** [comment]
> I met the similar problem in my project. It might due to the incompatible `org.apache.httpcomponents.httpclient` runtime version. In Twilio SDK `httpclient` version is 4.5.12:compile, while in my project it's 4.0.1:runtime. I manually added the dependency then it works.
> 
> ```
> <dependency>
>      <groupId>org.apache.httpcomponents</groupId>
>      <artifactId>httpclient</artifactId>
>      <version>4.5.12</version>
> </dependency>
> ```

cooool!!I met the same error and soluted by your answer,thanks!!

**Comment by anwarxbones** [comment]
I'm running into this same issue.  What combination of twilio and httpclient versions works?

**Comment by yoovan** [comment]
> I'm running into this same issue. What combination of twilio and httpclient versions works?

My project used twilio:8.0.* and httpclient:4.5.12

**Comment by TARUNGo** [comment]
thanks @yungang-bgl your solution worked

**Comment by hakkinepally** [comment]
I am using the JDK 1.8 and Twilio 8.17.0 and Http Client 4.5.12 and I am still experiencing the same issue. Please help me.

**Comment by juanlmelo** [comment]
This is basically a version conflict between Twilio's `org.apache.httpcomponents` library and the ones specified in your project. 

Try analyzing the results of `mvn dependency:tree` and remove httpclient definitions or make the proper exceptions and it should work.

In my case, I removed this from my project and It got fixed.


        <dependency>
            <groupId>org.apache.httpcomponents</groupId>
            <artifactId>httpclient</artifactId>
            <version>4.5.5</version>
        </dependency>

**Comment by dipanshu91** [comment]
I'm getting same issue with v9.0.0-rc.4. Spring Boot v2.5.6. Not added any external dependency for httpclient. What should be fix for this?

**Comment by autumoswitzerland** [comment]
Same problem with Twilio 9.0.1 and apache-httpclient 4.5.12, OpenJDK 13 compiled with compiler compliance leve 1.8 to be compatible with Java release 8.

**Comment by childish-sambino** [comment]
@autumoswitzerland 
Please open a new issue as closed issues are not tracked as closely.

**Comment by autumoswitzerland** [comment]
Ok, but what is the solution?

**Comment by autumoswitzerland** [comment]
Sorry for the confusion; just an add-on information:

I think it's really just a dependency-issue - one has to combine the right libs. E.g. newest Twilio 9.0.1 with Apache httpclient 4.5.13 works fine. Just explicitly define both in your maven dependencies and remove older versions. The missing method signature is available within the httpclient 4.5.13 lib.
